### PR TITLE
Feat/3380 (backend) add an endpoint to allow clients to set arbitrary parameters

### PR DESCRIPTION
**Pull request by raphj**
#3380 

:warning: this is the first time I touch many things, be particularly attentive during the review of this PR.

## Checkpoints

**For developers**

- [x] If relevant, manual tests have been done to ensure the stability of the whole application and that the involved feature works (not relevant)
- [x] The original issue is up to date w.r.t the latest discussions and contains a short summary of the implemented solution
- [x] Automated tests covering the feature or the fix, have been written, deemed irrelevant (give the reason), or an issue has been created to implement the test (give the link)

**For code reviewers**

- [x] The code is clear enough
- [x] If there are FIXMEs in the code, related issues are mentioned in the FIXME
- [x] If there are TODOs, NOTEs or HACKs in code, the date and the developer initials are present

**For testers**

- [x] Manual, quality tests have been done (irrelevant)

### CODE REVIEW

File: backend/tracim_backend/lib/core/event.py [hunk 1]
```diff
@@ -162,16 +162,6 @@ def get_paginated_messages_for_user(
...
-    def get_messages_count(
-        self,
-        user_id: int,
-        read_status: ReadStatus,
-        event_types: typing.List[EventTypeDatabaseParameters] = None,
-    ) -> int:
-        return self._base_query(
-            user_id=user_id, event_types=event_types, read_status=read_status
-        ).count()
-
```
Comment (grignards) [inline backend/tracim_backend/lib/core/event.py:174]: Isn't this method used for an API (https://github.com/tracim/tracim/pull/3394)?

File: backend/tracim_backend/lib/core/userconfig.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+    def __init__(self, user: User, session: TracimSession):
+        self._user = user
+        self._session = session
+
+    def get_config(self) -> UserConfig:
+        query = self._session.query(UserConfig)
+        return query.filter(UserConfig.user_id == self._user.user_id).one()
+
+    def get_all_params(self) -> Dict:
```
Comment (grignards) [inline backend/tracim_backend/lib/core/userconfig.py:19]: If you just want to say "any dict", you can use `dict` directly. `typing.Dict` is mainly useful to more precisely type the dict's contents (eg. `Dict[str, Any]` in your case).

File: backend/tracim_backend/migration/versions/d6ce2c95006f_added_userconfig.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+    connection = op.get_bind()
+    session = Session(bind=connection)
+
+    users = session.query(TemporaryUser).all()
+
+    for user in users:
+        session.add(TemporaryUserConfig(user_id=user.user_id))
+
+    session.commit()
```
Comment (grignards) [inline backend/tracim_backend/migration/versions/d6ce2c95006f_added_userconfig.py:71]: You can also directly use the created Table:
```python
user_configs_table = op.create_table(...)
users_table = Table("users", metadata, sa.Column("user_id", sa.Integer()))
connection.execute(user_configs_table.insert().from_select(sa.select([users_table.c.user_id])))
```
which has the advantage to be more concise and quicker as all data stays in the database + the use of a bulk insert.

Comment (raphj) [reply backend/tracim_backend/migration/versions/d6ce2c95006f_added_userconfig.py:71]: Nice, thanks

File: backend/tracim_backend/lib/core/event.py (hunk 1 above)
Comment (raphj) [reply backend/tracim_backend/lib/core/event.py:174]: Yes, this is a mistake, I'll revert this

File: backend/tracim_backend/models/userconfig.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    """
+
+    __tablename__ = "user_configs"
+
+    user_id = Column(Integer, ForeignKey("users.user_id"), nullable=False, primary_key=True)
+    user = relationship("User", remote_side=[User.user_id])
+
+    fields = Column(JSON, nullable=False, default={})
+
+    def __init__(self, user: User, fields: Optional[JSON] = None):
```
[the remark this answers is not in the document]
Comment (raphj) [reply backend/tracim_backend/models/userconfig.py:26]: Indeed

File: backend/tracim_backend/lib/core/userconfig.py (hunk 2 above)
Comment (raphj) [reply backend/tracim_backend/lib/core/userconfig.py]: Indeed, thanks. I used `Dict[str, Any]`

File: backend/tracim_backend/lib/core/user.py [hunk 5]
```diff
@@ -972,9 +972,18 @@ def undelete(self, user: User, do_save=False):
...
         self._session.flush()
 
+        # INFO - R.J - 2020-08-03 - the newly created user's object
+        # does not have an id until the session is flushed.
+        # This id is needed to create its configuration.
+
+        if is_new_user:
+            self._session.add(UserConfig(user=user))
+            self._session.flush()
+
```
[the remark this answers is not in the document]
Comment (raphj) [reply backend/tracim_backend/lib/core/user.py]: Ok, after @grignards' investigation, the issue was caused by the presence of a manual init in the UserConfig class.

File: backend/tracim_backend/models/userconfig.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    This table stores parameters for Tracim clients (including the Tracim frontend)
+    """
+
+    __tablename__ = "user_configs"
+
+    user_id = Column(
+        Integer,
+        ForeignKey(
+            "users.user_id",
+            name="fk_userconfig_user_id_users",
```
Comment (grignards) [inline backend/tracim_backend/models/userconfig.py:23]: This shouldn't be needed as sqlalchemy uses the defined naming convention for those.

File: backend/tracim_backend/views/core_api/user_controller.py [hunk 7]
```diff
@@ -720,6 +724,29 @@ def open_message_stream(self, context, request: TracimRequest, hapic_data=None)
...
 
+    @hapic.with_api_doc(tags=[SWAGGER_TAG__USER_CONFIG_ENDPOINTS])
+    @check_right(has_personal_access)
+    @hapic.input_path(UserIdPathSchema())
+    @hapic.output_body(UserConfigSchema())
+    def get_user_config(self, context, request: TracimRequest, hapic_data: HapicData) -> Dict:
+        """
+        get all the configuration parameters for the given user
+        """
+        config_api = UserConfigApi(user=request.candidate_user, session=request.dbsession)
```
Comment (grignards) [inline backend/tracim_backend/views/core_api/user_controller.py:735]: This should be `request.current_user`, `candidate_user` is the not authenticated/validated user. It shouldn't have any effect here, only good practice to use `current_user`.

File: backend/tracim_backend/lib/core/userconfig.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+from tracim_backend.models.auth import User
+from tracim_backend.models.tracim_session import TracimSession
+from tracim_backend.models.userconfig import UserConfig
+
+
+class UserConfigApi:
+    """Api to query user configuration"""
+
+    def __init__(self, user: User, session: TracimSession):
```
Comment (grignards) [inline backend/tracim_backend/lib/core/userconfig.py:12]: Other Api classes name the `user` parameter as `current_user`, so I would propose to rename it for coherency:
```suggestion
    def __init__(self, current_user: User, session: TracimSession):
```

File: backend/tracim_backend/lib/core/userconfig.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+        self._user = user
```
Comment (grignards) [inline backend/tracim_backend/lib/core/userconfig.py:13]:
```suggestion
        self._current_user = current_user
```

File: backend/tracim_backend/lib/core/userconfig.py [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+        self._session = session
+
+    def get_config(self) -> UserConfig:
+        query = self._session.query(UserConfig)
+        return query.filter(UserConfig.user_id == self._user.user_id).one()
```
Comment (grignards) [inline backend/tracim_backend/lib/core/userconfig.py:18]:
```suggestion
        return query.filter(UserConfig.user_id == self._current_user.user_id).one()
```

File: backend/tracim_backend/views/core_api/user_controller.py (hunk 7 above)
Comment (raphj) [reply backend/tracim_backend/views/core_api/user_controller.py:735]: I thought `current_user` was the one that makes the request and `candidate_user` the one given in the URL? current_user would therefore be wrong, no?

File: backend/tracim_backend/lib/core/user.py [hunk 11]
```diff
@@ -1030,15 +1039,15 @@ def execute_created_user_actions(self, user: User) -> None:
                 if agenda_already_exist:
                     logger.warning(
                         self,
-                        "user {} is just created but his own agenda already exist !!".format(
+                        "user {} has just been created but their own agenda already exists !!".format(
```
Comment (raphj) [inline backend/tracim_backend/lib/core/user.py:1039]:
```suggestion
                        "user {} has just been created but their own agenda already exists".format(
```

File: backend/tracim_backend/migration/versions/d6ce2c95006f_added_userconfig.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        "user_configs",
+        sa.Column("user_id", sa.Integer(), nullable=False),
+        sa.Column("fields", sa.JSON(), nullable=False, default={}),
+        sa.PrimaryKeyConstraint("user_id", name=op.f("pk_userconfig")),
+        sa.ForeignKeyConstraint(
+            ["user_id"],
+            ["users.user_id"],
+            name=op.f("fk_userconfig_user_id_users"),
+            onupdate="CASCADE",
+            ondelete="CASCADE",
```
[the remark this answers is not in the document]
Comment (raphj) [reply backend/tracim_backend/migration/versions/d6ce2c95006f_added_userconfig.py:40]: What do you mean?

We have:

```
class UserConfig(DeclarativeBase):
    """
    User Config definition.
    This table stores parameters for Tracim clients (including the Tracim frontend)
    """

    __tablename__ = "user_configs"

    user_id = Column(
        Integer,
        ForeignKey(
            "users.user_id",
            onupdate="CASCADE",
            ondelete="CASCADE",
        ),
        nullable=False,
        primary_key=True,
    )
```

[the remark this answers is not in the document]
Comment (grignards) [reply backend/tracim_backend/migration/versions/d6ce2c95006f_added_userconfig.py:40]: This was an old comment.

File: backend/tracim_backend/views/core_api/user_controller.py (hunk 7 above)
Comment (grignards) [reply backend/tracim_backend/views/core_api/user_controller.py:735]: … you're right, sorry.

File: backend/tracim_backend/migration/versions/d6ce2c95006f_added_userconfig.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (raphj) [reply backend/tracim_backend/migration/versions/d6ce2c95006f_added_userconfig.py:40]: Ah, sorry

File: backend/tracim_backend/views/core_api/user_controller.py (hunk 7 above)
Comment (raphj) [reply backend/tracim_backend/views/core_api/user_controller.py:735]: No problem :-)

**Pull request merged**